include/pqxx/blob.hxx

```diff
@@ -0,0 +1,207 @@
+/* Binary Large Objects interface.
+ *
+ * Read or write large objects, stored in their own storage on the server.
+ *
+ * DO NOT INCLUDE THIS FILE DIRECTLY; include pqxx/largeobject instead.
+ *
+ * Copyright (c) 2000-2020, Jeroen T. Vermeulen.
+ *
+ * See COPYING for copyright license.  If you did not receive a file called
+ * COPYING with this source code, please notify the distributor of this
+ * mistake, or contact the author.
+ */
+#ifndef PQXX_H_BLOB
+#define PQXX_H_BLOB
+
+#include "pqxx/compiler-public.hxx"
+#include "pqxx/internal/compiler-internal-pre.hxx"
+
+#include <cstdint>
+
+#include "pqxx/dbtransaction.hxx"
+
+
+namespace pqxx
+{
+/** Binary large object.
+ *
+ * This is how you store data that may be too large for the @c BYTEA type.
+ * Access operations are similar to those for a file: you can read, write,
+ * query or set the current reading/writing position, and so on.
+ *
+ * These large objects live in their own storage on the server, indexed by an
+ * integer object identifier ("oid").
+ *
+ * Two @c blob objects may refer to the same actual large object in the
+ * database at the same time.  Each will have its own reading/writing position,
+ * but writes to the one will of course affect what the other sees.
+ */
+class PQXX_LIBEXPORT blob
+{
+public:
+  /// Create a new, empty large object.
+  /** You may optionally specify an oid for the new blob.  If you do, then
+   * the new object will have that oid -- or creation will fail if there
+   * already is an object with that oid.
+   */
+  [[nodiscard]] static oid create(dbtransaction &, oid = 0);
+
+  /// Delete a large object, or fail if it does not exist.
+  static void remove(dbtransaction &, oid);
+
+  /// Open blob for reading.  Any attempt to write to it will fail.
+  [[nodiscard]] static blob open_r(dbtransaction &, oid);
+  // Open blob for writing.  Any attempt to read from it will fail.
+  [[nodiscard]] static blob open_w(dbtransaction &, oid);
+  // Open blob for reading and/or writing.
+  [[nodiscard]] static blob open_rw(dbtransaction &, oid);
+
+  /// You can default-construct a blob, but it won't do anything useful.
+  /** Most operations on a default-constructed blob will throw @c usage_error.
+   */
+  blob() = default;
```

Comment:
Why make it default constructable if it leads to a bad state? Doesn't that also mean the rest of the code has to check for the bad state?

Reply:
Good question.  I hope I have a good answer.

Initially I didn't allow default construction.  It was move semantics that gave rise to the unusable state.  And I wanted to keep move constructors/assignments so that I could have a wealth of different ways to create a blob, without unnatural overloading of constructors.

(I'm not exactly convinced at this stage that the separate read/write/read-write "constructors" carry their weight, but that's another story.)

Once I had move semantics an an unusable state, it seemed to make sense to support a "null large object" for cases where you don't know until run time whether you need a large object.

Next I'm considering whether to support querying for this state.  Best way to do that might be for a blob to remember its oid; we could then have a `create()` function which just immediately returns a `blob` opened in read-write mode, so you can start writing to it right away.

Reply:
>Once I had move semantics an an unusable state, it seemed to make sense to support a "null large object" for cases where you don't know until run time whether you need a large object.

From a technical stand point the moved from state only needs to support destruction. I think it's also important that you can move assign to it. If it's free to make it a valid state then why not do so, but that doesn't seem to be the case here.

For late initialisation I often wrap in a `std::optional` -- that let's me, for example, declare in an outer scope and initialise in an inner (and is a lot better than old code that did the same thing with an `auto_ptr`). The alternative is to use an IILE to create it.

I think I'm generally OK in saying that doing anything unsupported to the moved from object is undefined behaviour, especially if doing otherwise requires a significant code investment, or might hurt performance in hot code.

Reply:
It's not particularly expensive to check for an invalid state.  It adds some code, but doesn't complicate things.  Certainly it won't cost much in terms of run-time performance — the extra data is in cache, the extra branches are highly predictable, the cost is amortised against network and disc I/O, and I expect compilers to eliminate a lot of the checks entirely.

Now, I really don't want to make any _promises_ about the behaviour of an invalid object.  Doing that would complicate things a bit.  But I figure this can help people debug problems with unwarranted use of `std::move()`, perhaps.

Reply:
My one misgiving about the new "named constructor" style I'm exploring, where you don't call a public constructor but you call a static factory which _returns_ an object, is that I can't do it everywhere.  It'd be nice to be able to apply it consistently to all classes with a complex variety of construction variety... but it does require move semantics.

And unfortunately I have a few cases where I can't currently have move semantics.  It'd be great to have factory functions in `stream_from`, but it (or really, its base class `transactionfocus`) registers a pointer back to itself with the transaction.  That pointer is of course no longer valid when I move a `stream_from`.  I'll make a note to see if I can change that.

Reply:
Guaranteed RVO provides for move elision so it is possible to have a function that returns a non-copyable non-movable type because the return value is constructed on the caller's stack frame.

Reply:
Whoa, you're right of course!  That's going to be useful.  Down with crowded constructor overloads.

As for `blob`, there is actually a third scenario that gives rise to the unusable state.  You can `close()` a blob explicitly.  Not sure that's really useful, but it can fail, and the destructor won't be able to throw an exception.

Reply:
Here's the silly thing.  When I toyed with the idea of move semantics in the 1990s, I thought we might call it `operator return`.  So RVO should have occurred to me!